src/App/Fossa/Reachability/Upload.hs

```diff
@@ -0,0 +1,157 @@
+uploadReachability unit = case callGraphAnalysis unit of
+  NoCallGraphAnalysis -> pure unit
+  JarAnalysis [] -> pure unit
+  JarAnalysis someJars -> do
+    updatedJars <- traverse uploadJarAnalysis someJars
+    pure $ unit{callGraphAnalysis = JarAnalysis updatedJars}
+
+uploadJarAnalysis ::
+  ( Has FossaApiClient sig m
+  ) =>
+  ParsedJar ->
+  m ParsedJar
+uploadJarAnalysis jar = case parsedJarContent jar of
+  ContentStoreKey _ -> pure jar
+  ContentRaw bs -> do
+    key <- uploadContentForReachability bs
+    pure $ jar{parsedJarContent = ContentStoreKey key}
+
+callGraphOf ::
+  ( Has Logger sig m
+  , Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has Exec sig m
+  , Has (Lift IO) sig m
+  ) =>
+  DiscoveredProjectScan ->
+  m (Maybe SourceUnitReachability)
+callGraphOf (Scanned dpi (Success _ projectResult)) = do
+  let srcUnit = projectToSourceUnit False projectResult
+  let dependencies = dependenciesOf srcUnit
+  let displayId = sourceUnitType srcUnit <> "@" <> sourceUnitManifest srcUnit
+  let unit =
+        SourceUnitReachability
+          { srcUnitType = sourceUnitType srcUnit
+          , srcUnitManifest = sourceUnitManifest srcUnit
+          , srcUnitName = sourceUnitName srcUnit
+          , srcUnitOriginPaths = sourceUnitOriginPaths srcUnit
+          , srcUnitDependencies = dependencies
+          , callGraphAnalysis = NoCallGraphAnalysis
+          }
+  case (projectResultGraphBreadth projectResult, dpiProjectType dpi) of
+    -- if we do not have complete graph, i.e.. missing transitive dependencies
+    -- it is impossible to perform reachability, as we may not have all symbols
+    -- used in the application to perform accurate analysis
+    (Partial, _) -> do
+      logInfo . pretty $ "we do not support reachability analysis, with partial dependencies graph (skipping: " <> displayId <> ")"
+      pure Nothing
+    (Complete, MavenProjectType) -> context "maven" $ do
+      logDebug . pretty $ "trying to infer build jars from maven project: " <> show (projectResultPath projectResult)
+      analysis <- mavenJarCallGraph (projectResultPath projectResult)
+      pure . Just $ unit{callGraphAnalysis = analysis}
+    -- Exclude units for package manager/language we cannot support yet!
+    _ -> do
+      logInfo . pretty $ "we do not support reachability analysis for: " <> displayId <> " yet. (skipping)"
+      pure Nothing
+-- Not possible to perform reachability analysis for projects
+-- which were not scanned (skipped due to filter), as we do not
+-- complete dependency graph for them
+callGraphOf _ = pure Nothing
+
+-- | Unique locators from SourceUnit
+dependenciesOf :: SourceUnit -> [Locator]
+dependenciesOf srcUnit = maybe [] allLocators (sourceUnitBuild srcUnit)
+
+-- | Unique locators from SourceUnitBuild
+allLocators :: SourceUnitBuild -> [Locator]
+allLocators unit =
+  nub $
+    buildImports unit
+      ++ concatMap (\ud -> sourceDepLocator ud : sourceDepImports ud) (buildDependencies unit)
```

Comment:
[Prefer]: `++ concatMap (\ud -> sourceDepLocator ud : sourceDepImports ud) $ buildDependencies unit`

Reply:
This ran into issue prefix order in Haskell, going with what I originally had